File: src/main/java/com/moshi/srv/v1/im/ImController.java
package com.moshi.srv.v1.im;

import com.jfinal.aop.Before;
import com.jfinal.kit.HashKit;
import com.jfinal.kit.Ret;
import com.moshi.common.controller.BaseController;
import com.moshi.common.interceptor.UnlockInterceptor;
import com.moshi.common.plugin.Letture;
import io.jboot.Jboot;
import io.lettuce.core.SetArgs;

import java.util.Date;

public class ImController extends BaseController {
  @Before(UnlockInterceptor.class)
  public void index() {
    String key = HashKit.sha256(new Date().getTime() + "" + Math.random()).substring(0,12);
    Letture.sync().set(key, getLoginAccount(), SetArgs.Builder.ex(60 * 5));
    renderJson(
        Ret.ok("key", key)
            .set(
                "url",
                Jboot.configValue("host").trim()
                    + ":"
                    + Jboot.configValue("socketio.port").trim()
                    + "/srv/v1/im"));
  }
}

File: src/main/java/com/moshi/im/ImController.kt
package com.moshi.im

import com.jfinal.kit.Ret
import com.jfinal.kit.StrKit
import com.moshi.common.controller.BaseController
import com.moshi.common.plugin.Letture
import io.jboot.web.controller.annotation.RequestMapping
import io.reactivex.Observable
import java.util.concurrent.TimeUnit

@RequestMapping("/im")
class ImController : BaseController() {

  fun token() {
    if (isLogin) {
      val sync = Letture.sync()
      val token = StrKit.getRandomUUID()
      sync.sadd(IM_AUTH + loginAccountId, token)
      renderJson(Ret.ok("token", token))
      Observable.timer(5, TimeUnit.SECONDS).subscribe {
        sync.srem(IM_AUTH + loginAccountId, token)
      }
    } else {
      renderJson(Ret.fail("msg", "未登录"))
    }
  }

  companion object {
    val IM_AUTH = "moshi:im:auth:"
  }
}
